### PR TITLE
Implement IO metrics in VirtualFile

### PR DESCRIPTION
**Pull request by knizhnik**

### CODE REVIEW

File: pageserver/src/virtual_file.rs [hunk 1]
```diff
@@ -145,7 +184,9 @@ impl OpenFiles {
         // old file.
         //
         if let Some(old_file) = slot_guard.file.take() {
-            drop(old_file);
+            STORAGE_IO_TIME
+                .with_label_values(&["close", tenantid, timelineid])
+                .observe_closure_duration(|| drop(old_file));
         }
```
Comment (hlinnaka) [inline pageserver/src/virtual_file.rs:190]: This accounts the close-time to the *new* tenant and timeline that we are opening, not the file that we are closing.

I wouldn't expect closing files to account for much time, so maybe don't bother tracking the tenant and timeline for closing, and just collect one global metric for that.

Comment (knizhnik) [reply pageserver/src/virtual_file.rs]: Sorry, close statistics is really accounts in wrong way.
In principle it is possible to store path together with file or use filepath crate to obtain path for file...
But I agree with you: close is really fast - it is statistic from my notebook:
```
 pageserver_io_time_bucket{operation="close",tenant_id="0a99d78d5aec440ea44517492d8eef45",timeline_id="02d526fcffbe0a985853c8aaf188d2c6",le="0.000001"} 2373
pageserver_io_time_bucket{operation="close",tenant_id="0a99d78d5aec440ea44517492d8eef45",timeline_id="02d526fcffbe0a985853c8aaf188d2c6",le="0.00001"} 4394
pageserver_io_time_bucket{operation="close",tenant_id="0a99d78d5aec440ea44517492d8eef45",timeline_id="02d526fcffbe0a985853c8aaf188d2c6",le="0.0001"} 4487
```
so of close() calls are completed in less than 10 usec. 
I agree that grouping of close metrics by tenant/timelineid may be not needed.
At least for evicted files.I left them for files dropped by user, at least to be able to compare number of manually closed files from evicted files.

Comment (hlinnaka) [reply pageserver/src/virtual_file.rs]: Ok. A brief code comment to note that we don't track the tenant/timeline for evictions and why would be nice.

**Review by hlinnaka — APPROVED** [review]

**Comment by hlinnaka** [comment]
How much CPU overhead this adds, BTW? I think this now needs to calculate a hash of the tenant and timeline ID for every operation, in the `with_label_values` function. That's pretty cheap but not free. I'm reminded of the performance issue that lead to commit 579b5ee944f7bc36fe79e4a957b92a41d8dfedfa. That was about converting tenant/timeline IDs to strings rather than the hashing, but I wonder if even the hashing is noticeable? If it is, we could use `get_metric_with_label_values` here too, when a VirtualFile is opened.

**Comment by hlinnaka** [comment]
`get_metric_with_label_values` might actually be nicer than storing the tenant/timeline ids as strings anyway, even if there's no meaningful performance difference.

**Comment by knizhnik** [comment]
> How much CPU overhead this adds, BTW? I think this now needs to calculate a hash of the tenant and timeline ID for every operation, in the `with_label_values` function. That's pretty cheap but not free. I'm reminded of the performance issue that lead to commit [579b5ee](https://github.com/zenithdb/zenith/commit/579b5ee944f7bc36fe79e4a957b92a41d8dfedfa). That was about converting tenant/timeline IDs to strings rather than the hashing, but I wonder if even the hashing is noticeable? If it is, we could use `get_metric_with_label_values` here too, when a VirtualFile is opened.

I have also concerned about performance, especially with my previous implementation which has to perform tenantid/timelineid to hex string conversion and each call. Performance of test_pgbench is decreased from 4400 to 4200 TPS at my notebook, although dispersion is large (+-100).

In this implementation there is no Zid to hex conversion at all: them are extract from file path. Performance is about 4300 (but also with large dispersion).
So I can not precisely estimate overhead of maintaining IO metrics.

**Pull request merged**